store/tikv/coprocessor.go

```diff
@@ -107,35 +109,161 @@ const (
 // copTask contains a related Region and KeyRange for a kv.Request.
 type copTask struct {
 	region *Region
-	ranges []kv.KeyRange
+	ranges *copRanges
 
 	status   int
 	idx      int // Index of task in the tasks slice.
 	respChan chan *coprocessor.Response
 }
 
-func (t *copTask) pbRanges() []*coprocessor.KeyRange {
-	ranges := make([]*coprocessor.KeyRange, 0, len(t.ranges))
-	for _, r := range t.ranges {
+// copRanges is like []kv.KeyRange, but may has extra elements at head/tail.
+// It's for avoiding alloc big slice during build copTask.
+type copRanges struct {
+	first *kv.KeyRange
+	mid   []kv.KeyRange
+	last  *kv.KeyRange
+}
+
+func (r *copRanges) String() string {
+	var s string
+	r.do(func(ran *kv.KeyRange) {
+		s += fmt.Sprintf("[%q, %q]", ran.StartKey, ran.EndKey)
+	})
+	return s
+}
+
+func (r *copRanges) len() int {
+	var l int
+	if r.first != nil {
+		l += 1
+	}
+	l += len(r.mid)
+	if r.last != nil {
+		l += 1
+	}
+	return l
+}
+
+func (r *copRanges) at(i int) kv.KeyRange {
+	if r.first != nil {
+		if i == 0 {
+			return *r.first
+		}
+		i--
+	}
+	if i < len(r.mid) {
+		return r.mid[i]
+	}
+	return *r.last
+}
+
+func (r *copRanges) slice(from, to int) *copRanges {
+	var ran copRanges
+	if r.first != nil {
+		if from == 0 && to > 0 {
+			ran.first = r.first
+		}
+		if from > 0 {
+			from--
+		}
+		if to > 0 {
+			to--
+		}
+	}
+	if to <= len(r.mid) {
+		ran.mid = r.mid[from:to]
+	} else {
+		ran.mid, ran.last = r.mid[from:], r.last
+	}
+	return &ran
+}
+
+func (r *copRanges) do(f func(ran *kv.KeyRange)) {
+	if r.first != nil {
+		f(r.first)
+	}
+	for _, ran := range r.mid {
+		f(&ran)
+	}
+	if r.last != nil {
+		f(r.last)
+	}
+}
+
+func (r *copRanges) toPBRanges() []*coprocessor.KeyRange {
+	ranges := make([]*coprocessor.KeyRange, 0, r.len())
+	r.do(func(ran *kv.KeyRange) {
 		ranges = append(ranges, &coprocessor.KeyRange{
-			Start: r.StartKey,
-			End:   r.EndKey,
+			Start: ran.StartKey,
+			End:   ran.EndKey,
 		})
-	}
+	})
 	return ranges
 }
 
-func buildCopTasks(bo *Backoffer, cache *RegionCache, ranges []kv.KeyRange, desc bool) ([]*copTask, error) {
+func buildCopTasks(bo *Backoffer, cache *RegionCache, ranges *copRanges, desc bool) ([]*copTask, error) {
+	start := time.Now()
+	rangesLen := ranges.len()
+
 	var tasks []*copTask
-	for _, r := range ranges {
```

Comment:
this change looks complex...
closure is also bad for performance

what about simply change line 129 to `tasks := make([]*copTask, 0, len(ranges))`